lib/http3/server.c

```diff
@@ -430,6 +430,7 @@ static void set_state(struct st_h2o_http3_server_stream_t *stream, enum h2o_http
         assert(conn->delayed_streams.recv_body_blocked.prev == &stream->link || !"stream is not registered to the recv_body list?");
         break;
     case H2O_HTTP3_SERVER_STREAM_STATE_CLOSE_WAIT: {
+        stream->req.timestamps.response_end_at = h2o_gettimeofday(conn->super.ctx->loop);
```

Comment:
I'm afraid this is inconsistent from when we observe `response_end_at` in http1.c.

In http1.c, the moment is when all the data has been written to the TCP stack. The closest moment in http3 would be when all the data up to EOS has been written to QUIC packets (i.e. near the end of the `on_send_emit` function).

Transition to H2O_HTTP3_SERVER_STREAM_STATE_CLOSE_WAIT happens only after acks for all data have been received, hence is at least 1-RTT later than the desired moment. Also, the moment when receiving the last chunk in `do_send` is too early, as depending on the flow control and prioritization, that moment can be way earlier than when all the data gets sent.

Reply:
This makes sense now, thanks.

I believe to have addressed this in: https://github.com/h2o/h2o/pull/2973/commits/6c4efc8e1d66dbda6398a762a35b37b8b67d9f20